src/components/application_manager/src/commands/button_notification_to_mobile.cc

```diff
@@ -0,0 +1,172 @@
+  if (false ==
+      (*message_)[app_mngr::strings::msg_params].keyExists(
+          hmi_response::custom_button_id)) {
+    LOG4CXX_ERROR(
+        logger_, "CUSTOM_BUTTON mobile notification without custom_button_id.");
+    return;
+  }
+
+  if (!app) {
+    LOG4CXX_ERROR(logger_, "Application doesn't exist.");
+    return;
+  }
+
+  uint32_t custom_btn_id = 0;
+  custom_btn_id =
+      (*message_)[strings::msg_params][hmi_response::custom_button_id].asUInt();
+
+  if (false == app->IsSubscribedToSoftButton(custom_btn_id)) {
+    LOG4CXX_ERROR(logger_,
+                  "Application doesn't subscribed to this custom_button_id.");
+    return;
+  }
+
+  if ((mobile_api::HMILevel::HMI_NONE == app->hmi_level())) {
+    LOG4CXX_WARN(logger_,
+                 "CUSTOM_BUTTON mobile notification is allowed only "
+                     << "in FULL, LIMITED or BACKGROUND hmi level");
+    return;
+  }
+
+  SendButtonNotification(app);
+  return;
+}
+
+void ButtonNotificationToMobile::HandleOKButton(
+    app_mngr::ApplicationSharedPtr app) {
+  LOG4CXX_AUTO_TRACE(logger_);
+  using namespace application_manager;
+  LOG4CXX_DEBUG(logger_, "OK button received");
+
+  const auto subscribed_apps = SubscribedApps();
+  auto app_ptr = subscribed_apps.end();
+  if (app) {
+    app_ptr = std::find_if(subscribed_apps.begin(),
+                           subscribed_apps.end(),
+                           [&app](const ApplicationSharedPtr subscribed_app) {
+                             return app->app_id() == subscribed_app->app_id();
+                           });
+    if (app_ptr != subscribed_apps.end()) {
+      LOG4CXX_DEBUG(
+          logger_,
+          "Sending button press for this app  id: " << (*app_ptr)->app_id());
+      SendButtonNotification(*app_ptr);
+      return;
+    }
+  }
+
+  app_ptr = std::find_if(subscribed_apps.begin(),
+                         subscribed_apps.end(),
+                         [](const ApplicationSharedPtr subscribed_app) {
+                           return subscribed_app->IsFullscreen();
+                         });
+
+  if (app_ptr != subscribed_apps.end()) {
+    LOG4CXX_DEBUG(logger_,
+                  "Sending button press for app in FULL with app  id: "
+                      << (*app_ptr)->app_id());
+    SendButtonNotification(*app_ptr);
+    return;
+  }
+}
+
+const std::vector<ApplicationSharedPtr>
+ButtonNotificationToMobile::SubscribedApps() const {
+  const uint32_t btn_id = static_cast<uint32_t>(
+      (*message_)[strings::msg_params][hmi_response::button_name].asInt());
+  return application_manager_.applications_by_button(btn_id);
+}
+
+void ButtonNotificationToMobile::HandleMediaButton() {
+  LOG4CXX_AUTO_TRACE(logger_);
+  using namespace application_manager;
+  const auto subscribed_apps = SubscribedApps();
+
+  std::vector<ApplicationSharedPtr>::const_iterator it =
+      subscribed_apps.begin();
+  for (; subscribed_apps.end() != it; ++it) {
```

Comment:
@mked-luxoft Range based loop?

Reply:
@AByzhynar fixed in 2b001c6014447a350d41a2bcd1e834fd2c83057a